File: app/src/main/java/com/framgia/movieapp/utils/ObservableUtils.java
package com.framgia.movieapp.utils;

import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by FRAMGIA\pham.duc.nam on 23/06/2017.
 */

public class ObservableUtils {

    public static <T> Observable.Transformer<T, T> applyAsyncSchedulers() {
        return new Observable.Transformer<T, T>() {
            @Override
            public Observable<T> call(Observable<T> observable) {
                return observable.subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread());
            }
        };
    }

    public static <T> Observable.Transformer<T, T> applySyncSchedulers() {
        return new Observable.Transformer<T, T>() {
            @Override
            public Observable<T> call(Observable<T> observable) {
                return observable.subscribeOn(Schedulers.immediate())
                        .observeOn(Schedulers.immediate());
            }
        };
    }
}

File: app/src/main/java/com/framgia/movieapp/utils/ObservableUtils.kt
package com.framgia.movieapp.utils

import rx.Observable
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers

/**
 * Created by FRAMGIA\pham.duc.nam on 23/06/2017.
 */

object ObservableUtils {

  fun <T> applyAsyncSchedulers(): Observable.Transformer<T, T> {
    return Observable.Transformer<T, T> { observable ->
      observable.subscribeOn(Schedulers.io())
          .observeOn(AndroidSchedulers.mainThread())
    }
  }

  fun <T> applySyncSchedulers(): Observable.Transformer<T, T> {
    return Observable.Transformer<T, T> { observable ->
      observable.subscribeOn(Schedulers.immediate())
          .observeOn(Schedulers.immediate())
    }
  }
}
